Unary plus operator (Issue #16073)

wuisawesome opened the pull request:
- [x] closes #16073
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

Continuation from https://github.com/pandas-dev/pandas/pull/16103.

wuisawesome committed:
ENH: Unary plus operator, deals with issue #16073

Overrides `+` operator, providing consistent behavior with `-`.
Creates tests for the new operator for series and dataframes.
Edits "What's new" to reflect the new feature

wuisawesome committed:
ENH: Unary plus operator, deals with issue #16073

Overrides `+` operator, providing consistent behavior with `-`.
Creates tests for the new operator for series and dataframes.

wuisawesome committed:
ENH: Unary plus operator, deals with issue #16073

Overrides `+` operator, providing consistent behavior with `-`.
Creates tests for the new operator for series and dataframes.
Edits "What's new" to reflect the new feature

wuisawesome committed:
What's new

wuisawesome committed:
Merge branch 'unary_plus_operator' of github.com:wuisawesome/pandas into unary_plus_operator

jreback commented on pandas/tests/series/test_operators.py:
```diff
@@ -91,6 +91,9 @@ def check(series, other, check_reverse=False):
         check(tm.makeFloatSeries(), tm.makeFloatSeries(), check_reverse=True)
 
+    def test_pos(self):
+       assert_series_equal(+self.series, 1 * self.series)  #Ensures no rounding issues
```
comment is not necessary

jreback commented on pandas/tests/series/test_operators.py:
```diff
@@ -91,6 +91,9 @@ def check(series, other, check_reverse=False):
 
+    def test_pos(self):
+       assert_series_equal(+self.series, 1 * self.series)  #Ensures no rounding issues
+
```
add a couple of raise assertion on strings

jreback replied:
this would technically work on datetimes, but I think we should disallow this (pls add tests for ``__neg__`` as well).

jreback replied:
e.g. 

```
In [9]: -Series(list('abc'))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-9-93660e79330e> in <module>()
----> 1 -Series(list('abc'))

/Users/jreback/pandas/pandas/core/generic.py in __neg__(self)
    865             arr = operator.inv(values)
    866         else:
--> 867             arr = operator.neg(values)
    868         return self.__array_wrap__(arr)
    869 

TypeError: bad operand type for unary -: 'str'
```

jreback replied:
```
In [10]: Series(pd.date_range('20130101', periods=3))
Out[10]: 
0   2013-01-01
1   2013-01-02
2   2013-01-03
dtype: datetime64[ns]

In [11]: -Series(pd.date_range('20130101', periods=3))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-11-23bfb90f61eb> in <module>()
----> 1 -Series(pd.date_range('20130101', periods=3))

/Users/jreback/pandas/pandas/core/generic.py in __neg__(self)
    865             arr = operator.inv(values)
    866         else:
--> 867             arr = operator.neg(values)
    868         return self.__array_wrap__(arr)
    869 

TypeError: ufunc 'negative' did not contain a loop with signature matching types dtype('<M8[ns]') dtype('<M8[ns]')

In [12]: +Series(pd.date_range('20130101', periods=3))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-12-b72a11ca274c> in <module>()
----> 1 +Series(pd.date_range('20130101', periods=3))

TypeError: bad operand type for unary +: 'Series'

```

I want [12] still to raise after this change.

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -1196,6 +1195,8 @@ Other API Changes
 - ``DataFrame`` and ``Panel`` constructors with invalid input will now raise ``ValueError`` rather than ``pandas.core.common.PandasError``, if called with scalar inputs and not axes; The exception ``PandasError`` is removed as well. (:issue:`15541`)
 - The exception ``pandas.core.common.AmbiguousIndexError`` is removed as it is not referenced (:issue:`15541`)
 
+- `Series` and `DataFrame` now have unary plus operators (:issue:`16073`). The plus operator behaves consistently with the unary minus operator. 
```
use double backticks here around Series, DataFrame. Put the issue reference at the end.

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -213,8 +213,7 @@ See :ref:`the docs here <io.pickle.compression>`
 
 Using an explicit compression type
 
-.. ipython:: python
```
don't change other things in the file

wuisawesome replied:
I'm not certain why this happened. The rest of the diff shows that the same identical line is added again, thus making no actual changes to the file

wuisawesome committed:
Operator edge cases, better tests

Updated plus operator edge case. Datatype that implement + but not -
should not be able to use + to maintain symmetry

Rewrites tests to work better with strings and other non numeric
datatypes that implement `*` operator

wuisawesome committed:
Merge remote-tracking branch 'pandas-dev/master' into unary_plus_operator

wuisawesome committed:
Added double ticks to whats new

wuisawesome committed:
Merging

wuisawesome committed:
Merge remote-tracking branch 'pandas-dev/master' into unary_plus_operator

wuisawesome commented:
Hopefully I've addressed all your concerns and rebased.

jreback commented on pandas/tests/frame/test_operators.py:
```diff
@@ -343,9 +343,15 @@ def test_logical_with_nas(self):
 
+    def test_pos(self):
+        try:
+            assert_frame_equal(+self.frame, self.frame.apply(lambda x: +self.frame))
```
dont' test like this. instead use a couple of sample frames that have singular dtypes (can be a single column). test the valid ones via ``assert_frame_equal(-frame, -1 * frame)``. The invalid ones should raise.

jreback commented on pandas/tests/series/test_operators.py:
```diff
@@ -91,8 +91,15 @@ def check(series, other, check_reverse=False):
+       try:
+           assert_series_equal(+self.series, self.series.apply(lambda x: +x))
+       except TypeError:
+           if not (len(self.frame) > 0 and isinstance(self.frame[0], pd.Timestampe)):
```
same

jreback commented on pandas/core/generic.py:
```diff
@@ -859,6 +859,19 @@ def _indexed_same(self, other):
 
+    def __pos__(self):
+        values = _values_from_object(self)
+        if values.dtype == np.bool_:
```
something like:
```
if is_bool_dtype(values):
   ...
elif not (is_numeric_dtype(values) or is_timedelta64_dtype(values)):
   raise
```

jreback replied:
add the same logic to ``__neg__``

jreback commented:
can you rebase / update

jreback commented:
can you move whatsnew & rebase / update

jreback commented:
closing as stale. if you want to re-visit, pls comment and we can re-open.
